### PR TITLE
Azure DevOps CICD functionality

### PR DESCRIPTION
**Pull request by niall-turbitt**
## Description

* Functionality added to MLOps Stacks to enable selection of CICD platform. This PR now has an additional cookiecutter prompt `cicd_platform` at project instantiation
* Ability to select either `cicd_platform = gitHub` or `cicd_platform = azureDevOpsServices`
* GitHub Actions CICD workflow replicated for Azure DevOps
* The Terraform CICD bootstrap when Azure DevOps (`azureDevOpsServices`) is selected as the CICD platform will do the following:
    1. Configure SP for staging and prod (as is done in the GHA workflow)
    2. Set up git for staging and prod workspaces (as is done in the GHA workflow)
    3. Assign the SP suitable permissions such that they have access to the storage backend
    4. Create a variable group in Azure DevOps Pipelines to propagate variables needed in CICD pipelines
    5. Create two Azure Pipelines (`tests-ci.yml` and `terraform-cicd.yml`)
    6. Configure a Build validation policy to enable PR validation against the default branch
* Notably, in the Azure DevOps implementation we remove the step in the GHA workflow where the TrainModelDeploy notebook triggers model deployment through a trigger to GitHub. In the ADO implementation, the model is automatically pushed to the Model Registry stage contingent on the env the job is triggered in (i.e if the model train job is run in staging workspace, the model is automatically pushed to staging, is run in production workspace, automatically register to stage=production)   

## Outstanding issues
- [ ] cookiecutter/jinja templating needs to be validated. Currently, the `cicd_platform` will always be `azureDevOpsServices` as it is the first option listed in `cookiecutter.json`
- [ ] CICD workflow for Azure DevOps needs to be validated
- [ ] Issue populating link correctly once the project is instantiated [here](https://github.com/niall-turbitt/mlops-stack/blob/niall-turbitt-ado-refactor/%7B%7Bcookiecutter.project_name%7D%7D/docs/ml-pull-request.md?plain=1#L22)
- [ ] Local testing - currently 2 tests failing when running `pytest test` 

**Additional Note**

* Currently the service principals produced [here](https://github.com/niall-turbitt/mlops-stack/blob/niall-turbitt-ado-refactor/%7B%7Bcookiecutter.project_name%7D%7D/.mlops-setup-scripts/cicd/main-azure.tf#L11) do not work automatically with the Azure DevOps workflow
* i.e Using the application ID and client secrets for staging and prod produced by [this module](https://registry.terraform.io/modules/databricks/mlops-azure-project-with-sp-creation/databricks/latest)
* The AAD application for those two app IDs is been created when the CICD bootstrap is executed, but enterprise applications are not been created for them.
* The current workaround is to manually assign these SP the correct enterprise application permissions through the Azure admin console after running `./mlops-setup-scripts/cicd/bootstrap.py`

### CODE REVIEW

**Comment by niall-turbitt** [comment]
**Additional Note**
* Currently the service principals produced [here](https://github.com/niall-turbitt/mlops-stack/blob/niall-turbitt-ado-refactor/%7B%7Bcookiecutter.project_name%7D%7D/.mlops-setup-scripts/cicd/main-azure.tf#L11) do not work automatically with the Azure DevOps workflow
* i.e Using the application ID and client secrets for staging and prod produced by [this module](https://registry.terraform.io/modules/databricks/mlops-azure-project-with-sp-creation/databricks/latest)
* The AAD application for those two app IDs is been created when the CICD bootstrap is executed, but enterprise applications are not been created for them. 
* The current workaround is to manually assign these SP the correct enterprise application permissions through the Azure admin console after running `./mlops-setup-scripts/cicd/bootstrap.py`

File: cookiecutter.json [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
   "project_name": "my-mlops-project",
   "cloud": ["azure", "aws"],
+  "cicd_platform": ["azureDevOpsServices", "gitHub"],
```
Comment (vladimirk-db) [inline cookiecutter.json:4]: Why not have the existing selection (gitHub) first, as the default one? That way by default it preserves existing behavior.

File: tests/example-project-configs/azure-github.yaml [hunk 2]
```diff
@@ -1,6 +1,8 @@
+# Example project config to use to preview stack changes for Azure, using GitHub for CICD
```
[the remark this answers is not in the document]
Comment (vladimirk-db) [reply tests/example-project-configs/azure-github.yaml:1]: Also, other places, as well, for consistency.

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        DATABRICKS_HOST: $(STAGING_DATABRICKS_HOST)
+        DATABRICKS_TOKEN: $(DATABRICKS_TOKEN)
+      displayName: 'Trigger integration test'
+
+    - task: PublishTestResults@2
+      condition: succeededOrFailed()
+      inputs:
+        testResultsFormat: 'JUnit'
+        testResultsFiles: '**/test-*.xml'
+        failTaskOnFailedTests: true
```
Comment (vladimirk-db) [inline {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml:130]: I wonder if there's any easy way to dedupe this job config? I think it's the same as the one we use for the github actions integration test. @smurching

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 4]
```diff
@@ -112,13 +128,20 @@ on the current repo.
...
 2. Call back from
    Databricks -> GitHub Actions to trigger a model deployment deployment workflow when
    automated model retraining completes, i.e. perform step (2) in
    [this diagram](https://github.com/databricks/mlops-stack/blob/main/Pipeline.md#model-training-pipeline).
 
 If using GitHub as your hosted Git provider, you can generate a Git token through the [token UI](https://github.com/settings/tokens/new);
 be sure to generate a token with "Repo" scope. If you have SSO enabled with your Git provider, be sure to authorize your token.
+{% elif cookiecutter.cicd_platform == "azureDevOpsServices" -%}
+
+If using AzureDevOps as your CI/CD platform, you can generate a PAT token by following the steps described [here](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)
```
Comment (vladimirk-db) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:141]:
```suggestion
If using AzureDevOps as your CI/CD platform, you can generate a PAT by following the steps described [here](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)
```

File: hooks/post_gen_project.py [hunk 5]
```diff
@@ -1,12 +1,27 @@
...
+
+def remove_filepath(filepath):
+    if os.path.isfile(filepath):
+        os.remove(filepath)
+    elif os.path.isdir(filepath):
+        shutil.rmtree(filepath)
+
 
 project_name = "{{cookiecutter.project_name}}"
+print(f"project_name: {project_name}")
```
Comment (smurching) [inline hooks/post_gen_project.py:13]: Were these prints intentionally included? Or just for debugging (in which case we should remove them)

File: cookiecutter.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (smurching) [reply cookiecutter.json:4]: I think the strings here are what the Databricks REST API expects: https://docs.databricks.com/dev-tools/api/latest/repos.html#operation/create-repo.

To @vladimirk-db 's point, we could even do something like `["Azure DevOps", "GitHub Actions"]`? We can transform the user-provided string into the target CI/CD provider in our `pre_gen_project.py` hook as is currently done for the other parameters here. But ok with doing that in a follow-up PR

File: README.md [hunk 6]
```diff
@@ -46,6 +47,7 @@ To create a new project, run:
 This will prompt for parameters for project initialization. Some of these parameters are required to get started:
  * ``project_name``: name of the current project
  * ``cloud``: Cloud provider you use with Databricks (AWS, Azure, or GCP)
+ * ``cicd_platform`` : CI/CD platform of choice (azureDevOpsServices or gitHub)
```
[the remark this answers is not in the document]
Comment (smurching) [reply README.md:50]: Yeah, maybe we can use the human-facing names ("Azure DevOps" and "GitHub Actions") here

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+# 1. Service principal tenant ID
+# 2. Service principal application (client) ID
+# 3. Service principal client secret
+#
+# It saves the token as the DATABRICKS_TOKEN environment variable for use in subsequent GitHub workflow steps.
+DATABRICKS_TOKEN=$(curl --fail -X POST -H 'Content-Type: application/x-www-form-urlencoded' \
+                                              "https://login.microsoftonline.com/$1/oauth2/v2.0/token" \
+                                              -d "client_id=$2" \
+                                              -d 'grant_type=client_credentials' \
+                                              -d 'scope=2ff814a6-3304-4ab8-85cb-cd0e6f879c1d%2F.default' \
```
[the remark this answers is not in the document]
Comment (smurching) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh:13]: It maps to the "programmatic ID for Azure Databricks" described in https://learn.microsoft.com/en-us/azure/databricks/dev-tools/api/latest/aad/service-prin-aad-token#--get-an-azure-ad-access-token-with-the-microsoft-identity-platform-rest-api. We could make this an input argument to facilitate customization for Databricks-internal users, but for all customers this scope value should be correct. That said maybe in internal docs/troubleshooting wiki we can document the scope to provide for internal users (we shouldn't actually advertise that internal scope value in public docs)

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+      enabled: 'true'
+      env:
+        ARM_TENANT_ID: $(PROD_AZURE_SP_TENANT_ID)
+        ARM_CLIENT_ID: $(PROD_AZURE_SP_APPLICATION_ID)
+        ARM_CLIENT_SECRET: $(PROD_AZURE_SP_CLIENT_SECRET)
+        DATABRICKS_HOST: $(PROD_DATABRICKS_HOST)
+
+# Run Terraform CD stage after successfully merging to {{cookiecutter.default_branch}}
+- stage: TerraformCD
+  displayName: 'Terraform Deployment for mlops-stack-ado'
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml:165]: The pipeline name here probably should include the `{{cookiecutter.project_name}}` rather than `mlops-stack-ado`?

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    - script: |
+       terraform fmt -check
+      displayName: 'Terraform fmt'
+      continueOnError: true
+
+    - script: |
+        terraform init
+      displayName: 'Terraform Init'
+      workingDirectory: $(working-directory)/$(environment)
+      enabled: 'true'
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml:215]: Question: do we need to set `enabled: 'true'` on each step here / what's the purpose? I found https://learn.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops&tabs=schema#define-a-containers-resource but it seems unrelated

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml (hunk 3 above)
Comment (smurching) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml:130]: Indeed it is. We could pull it out into a common shell script and run it in GitHub Actions too. Longer term, our Jobs team will improve the story here for testing and deploying jobs (note that even if we pull this out into a shell script shared with GitHub Actions, the cluster spec used for the job is currently duplicated with what's in Terraform)

Comment (smurching) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml:130]: I don't think it's urgent to update our GitHub Actions YAMLs to use the Databricks CLI (we'd lose some of the readability benefits of using the native https://github.com/databricks/run-notebook Action to run integration tests), but it would be nice to pull the `databricks runs submit` invocation here into a shell script so we can reuse it across other CI/CD providers for which we don't have an analog for https://github.com/databricks/run-notebook

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 10]
```diff
@@ -112,13 +128,20 @@ on the current repo.
 This token is used to:
 1. Fetch ML code from the current repo to run on Databricks for CI/CD (e.g. to check out code from a PR branch and run it
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:129]: Nit: It's a bit weird to see this render as a single-item list:

image


Maybe we can just if/else the whole block so that it looks something like this:

```
The setup script prompts a Git token with both read and write permissions
on the current repo.

This token is used to fetch ML code from the current repo to run on Databricks for CI/CD (e.g. to check out code from a PR branch and run it
during CI/CD). You can generate a PAT token for Azure DevOps by following the steps described [here](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows).
```

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 11]
```diff
@@ -112,13 +128,20 @@ on the current repo.
...
    automated model retraining completes, i.e. perform step (2) in
    [this diagram](https://github.com/databricks/mlops-stack/blob/main/Pipeline.md#model-training-pipeline).
 
 If using GitHub as your hosted Git provider, you can generate a Git token through the [token UI](https://github.com/settings/tokens/new);
 be sure to generate a token with "Repo" scope. If you have SSO enabled with your Git provider, be sure to authorize your token.
+{% elif cookiecutter.cicd_platform == "azureDevOpsServices" -%}
+
+If using AzureDevOps as your CI/CD platform, you can generate a PAT token by following the steps described [here](https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)
+
+**IMPORTANT**: in order to run Git commands from a build pipeline you must [grant version control permissions](https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/git-commands?view=azure-devops&tabs=yaml#grant-version-control-permissions-to-the-build-service) to the build service.
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:143]: Just to confirm, at what stage in the setup process should people grant the permissions (seems like from the linked docs they need to do it as a prereq)? Maybe we can clarify that they should do it at the project level in their hosted Git repository/update the wording to indicate that it's not an optional step?

```suggestion
### Grant version control permissions to CI/CD
The provided CI/CD workflows attempt to run `git` commands to commit and modify files. To ensure the workflows can work properly, [grant version control permissions](https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/git-commands?view=azure-devops&tabs=yaml#grant-version-control-permissions-to-the-build-service) within your hosted Azure DevOps repo
```

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 12]
```diff
@@ -136,16 +159,27 @@ export AWS_REGION="us-east-1"
 {% endif -%}
 python .mlops-setup-scripts/terraform/bootstrap.py
 ```
+{% if cookiecutter.cicd_platform == "azureDevOpsServices" -%}
+This initial bootstrap will produce an ARM access key. This key is required as a variable in the next step. To view this
+key locally and copy the key for this step you can do the following `vi ~/.mlops-stack-ado-cicd-terraform-secrets.json`.
+{% endif -%}
 
-Then, fill in vars and run the command below to bootstrap CI/CD.
+Then, run the following command providing the required vars to bootstrap CI/CD.
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:167]:
```suggestion
Then, run the following command, providing the required vars to bootstrap CI/CD.
```

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 13]
```diff
@@ -136,16 +159,27 @@ export AWS_REGION="us-east-1"
 {% endif -%}
 python .mlops-setup-scripts/terraform/bootstrap.py
 ```
+{% if cookiecutter.cicd_platform == "azureDevOpsServices" -%}
+This initial bootstrap will produce an ARM access key. This key is required as a variable in the next step. To view this
+key locally and copy the key for this step you can do the following `vi ~/.mlops-stack-ado-cicd-terraform-secrets.json`.
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:164]:
```suggestion
key locally and copy the key for this step you can do the following `vi ~/.{{cookiecutter.project_name}}-cicd-terraform-secrets.json`.
```

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml (hunk 8 above)
Comment (smurching) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml:165]: Same comment for other instances of `mlops-stack-ado` in this PR (maybe we can find-and-replace with `{{cookiecutter.project_name}}`)

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/azure-devops.tf [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+data "azuredevops_project" "project" {
+  name = var.azure_devops_project_name
+}
+
+data "azuredevops_git_repository" "repo" {
+  project_id = data.azuredevops_project.project.project_id
+  name       = var.azure_devops_repo_name
+}
+
+// Required that service principals have access to storage backend. Assign permissions
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/azure-devops.tf:22]: Nice, glad this ended up working!

File: {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py [hunk 15]
```diff
@@ -65,3 +73,14 @@
     f"Successfully triggered model CD deployment workflow for {model_uri}. See your CD provider to check the "
     f"status of model deployment"
 )
+
+{% elif cookiecutter.cicd_platform == "azureDevOpsServices" -%}
+# COMMAND ----------
+from deploy import deploy
+deploy(model_uri, env)
```
Comment (smurching) [inline {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py:80]:
```suggestion
# TODO: Add any additional pre-deployment checks here to ensure model quality before calling
# `deploy` to push the model to the desired environment.
deploy(model_uri, env)
```

Maybe we can add a user-facing TODO here to make it clear where to add pre-deploy checks

File: {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py [hunk 16]
```diff
@@ -65,3 +73,14 @@
...
+
+# COMMAND ----------
+print(
+    f"Successfully triggered model deployment workflow for {model_uri}"
+)
+{% endif -%}
```
[the remark this answers is not in the document]
Comment (smurching) [reply {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py:86]: Good question, we discussed starting with this for now to unblock getting Azure DevOps working sooner - IIRC triggering an ADO pipeline via REST API (and saving the appropriate credentials for that in Databricks) was doable but somewhat complicated

[the remark this answers is not in the document]
Comment (smurching) [reply {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py:86]: I'll let @niall-turbitt add more detail

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (niall-turbitt) [reply README.md:50]: resolved

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh:13]: is there anything that needs to be actioned on this for now?

File: cookiecutter.json (hunk 1 above)
Comment (niall-turbitt) [reply cookiecutter.json]: agreed. This is an artefact of what I was referencing in the above "outstanding issues" with "Currently, the cicd_platform will always be azureDevOpsServices as it is the first option listed in cookiecutter.json"

I'm still running into this issue and can't figure out (at least when I'm running the tests outlined in the readme) why the first option of the cicd_platform arg is always being selected. For example, currently if I have ["gitHub", "azureDevOpsServices"] in line 4, cookiecutter will always populate with "gitHub", regardless of the test config used. @vladimirk-db Would you be able to check if you run into the same issue when running those tests?

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml (hunk 3 above)
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml:130]: I had actually started to try to do something similar to the above suggestion. I looked at pulling the common cluster config out, and have a `payload_cluster_config.json` with the configs for a new cluster, but it started to look a bit overly complex. Specifically for passing that json into the task in the Azure testing-ci.yml script

File: hooks/post_gen_project.py (hunk 5 above)
Comment (niall-turbitt) [reply hooks/post_gen_project.py:13]: yep a remnant of debugging, will remove

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml (hunk 8 above)
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/terraform-cicd.yml:165]: good catch, thought I had indeed done a find and replace on those!

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md (hunk 11 above)
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md]: great suggestion, have incorporated

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md (hunk 10 above)
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:130]: agreed, and fixed

File: {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (vladimirk-db) [reply {{cookiecutter.project_name}}/notebooks/TriggerModelDeploy.py]: Sounds good, to strip this for now. thanks for the context, @niall-turbitt.

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml (hunk 3 above)
Comment (vladimirk-db) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml:130]: I see - agreed, I think this would be best solved w/ "workflows as code". We can reference the (CI/CD provider-agnostic) workflow configs from GH Actions, ADO, Terraform and other places.

File: cookiecutter.json (hunk 1 above)
Comment (vladimirk-db) [reply cookiecutter.json]: Hmm not sure why the tests are not honoring selecting the non-default platform, can you post a link to the (failing/misbehaving) test run please? @smurching  would have more context. We should compare with the tests parameterizing "cloud" (azure/aws) since they would have had to overcome the same issue.

(My suggestion was to just swap the order in that cicd_platform options list, so that Github comes first to preserve existing default behavior.)

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (smurching) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh:13]: I think it's fine to leave as is for now

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml (hunk 3 above)
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml:130]: Agreed with Sid for supporting other git providers it would then make sense to pull out into a shell script. Until we have introduced those other providers I would vote to keep this as is for now

File: cookiecutter.json (hunk 1 above)
Comment (niall-turbitt) [reply cookiecutter.json]: the failure is resulting when I go to create the example projects as outlined in the [README](https://github.com/niall-turbitt/mlops-stack/blob/main/README.md#previewing-stack-changes). Regardless of the example config file specified, the resulting project will be whatever the first option of `cicd_provider` is.

For example, when the [cicd_platform options](https://github.com/niall-turbitt/mlops-stack/blob/48f04c36a7409f3504bf7568851de03aa2310351/cookiecutter.json#L4) are in the following order ["Azure DevOps", "GitHub Actions"] and I run the following:
`cookiecutter "$MLOPS_STACK_PATH" --config-file "$MLOPS_STACK_PATH/tests/example-project-configs/azure/azure-devops.yaml" --no-input --overwrite-if-exists`
the generated project will create the "Azure DevOps" version as expected. If the order of `cicd_platform` is reversed to ["GitHub Actions", "Azure DevOps"] the generated project will be the GitHub Actions version.

It does not honour the cicd choice in the example-project-config file specified, and will take the first option of the list for `cicd_platform` in `cookiecutter.json`. @smurching is this something you could validate?

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md [hunk 17]
```diff
@@ -16,6 +17,21 @@ to other CI/CD providers by running the same shell commands, with a few caveats:
   hardcodes the API endpoint for triggering a GitHub Actions workflow. Update `notebooks/TriggerModelDeploy.py`
   to instead hit the appropriate REST API endpoint for triggering model deployment CD for your CI/CD provider.
 
+{% elif cookiecutter.cicd_platform == "azureDevOpsServices" -%}
```
[the remark this answers is not in the document]
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/README.md:20]: @vladimirk-db I think we should have this covered with point vii. [here](https://github.com/niall-turbitt/mlops-stack/tree/niall-turbitt-ado-refactor/%7B%7Bcookiecutter.project_name%7D%7D/.mlops-setup-scripts#verify-permissions) :

"Verify that you can create and manage service principals in the AAD tenant, by opening the
       [App registrations UI](https://portal.azure.com/#view/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/~/RegisteredApps)
       under the Azure Active Directory resource within the Azure portal. Then, verify that you can click "New registration" to create
       a new AAD application, but don't actually create one. If unable to click "New registration", ask your Azure admin to grant you [Application.ReadWrite.All](https://docs.microsoft.com/en-us/graph/permissions-reference#application-resource-permissions) permissions"

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh [hunk 18]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+# This script generates an Azure Active Directory token for an Azure service principal,
```
[the remark this answers is not in the document]
Comment (niall-turbitt) [reply {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh:2]: likewise for the point below, I think [these](https://github.com/niall-turbitt/mlops-stack/tree/niall-turbitt-ado-refactor/%7B%7Bcookiecutter.project_name%7D%7D/.mlops-setup-scripts#verify-permissions) prerequisites are sufficient?

File: .gitignore [hunk 19]
```diff
@@ -13,10 +13,3 @@ terraform.rc
 __pycache__/
 .cache
 *.pyc
-
-# Exclude local environment directory
-venv
-# Exclude local distribution/packaging directories
```
Comment (smurching) [inline .gitignore:19]: @niall-turbitt were these changes necessary / can we undo them?

File: README.md [hunk 20]
```diff
@@ -9,7 +9,8 @@ The default stack in this repo includes three modular components:
...
-| CI/CD                      | [GitHub Actions](https://github.com/actions) workflows to test and deploy ML code and resources                                                                    | Ship ML code faster and with confidence: ensure all production changes are performed through automation and that only tested code is deployed to prod                                   |
+| CI/CD                      | One of the following: <ul><li>[GitHub Actions](https://github.com/actions) workflows to test and deploy ML code and resources</li><li>[Azure DevOps](https://azure.microsoft.com/en-gb/products/devops/) to test and deploy ML code and resources</ul>              | Ship ML code faster and with confidence: ensure all production changes are performed through automation and that only tested code is deployed to prod                                   |
```
Comment (smurching) [inline README.md:12]: Super minor nit, maybe here we could just say "GitHub Actions or Azure DevOps workflows to test and deploy ML code and resources ...", instead of switching to a bulleted list, just to save a bit of space/format a bit more nicely

File: {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+# given the following positional command-line arguments:
+# 1. Service principal tenant ID
+# 2. Service principal application (client) ID
+# 3. Service principal client secret
+#
+# It saves the token as the DATABRICKS_TOKEN environment variable for use in subsequent GitHub workflow steps.
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.azure/devops-pipelines/scripts/generate-aad-token.sh:8]:
```suggestion
# It saves the token as the DATABRICKS_TOKEN environment variable for use in subsequent workflow steps.
```

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/main-azure.tf [hunk 22]
```diff
@@ -38,34 +38,77 @@ provider "databricks" {
...
+  owners       = [data.azuread_client_config.current.object_id]
+}
+
+resource "azuread_service_principal" "staging_service_principal" {
+  application_id               = module.azure_create_sp.staging_service_principal_application_id
+  app_role_assignment_required = false
+  owners                       = [data.azuread_client_config.current.object_id]
+
+  feature_tags {
+    enterprise = true
```
Comment (smurching) [inline {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/main-azure.tf:88]: Out of curiosity, do these actually need to be enterprise applications / do we want to be opinionated about that? Or could we keep things simple/have them just be linked Azure AD service principals to start with?

**Review by smurching — APPROVED** [review]
@niall-turbitt I left a few last comments, but this LGTM once they're addressed :)

File: README.md (hunk 20 above)
Comment (niall-turbitt) [reply README.md:12]: yea agreed on this actually. fixed

File: {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/main-azure.tf (hunk 22 above)
Comment (alexott) [reply {{cookiecutter.project_name}}/.mlops-setup-scripts/cicd/main-azure.tf:88]: service principal == enterprise application. Only with them you can assign permissions, etc. Otherwise it's just an application, not service principal
